packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php

```diff
@@ -116,4 +130,22 @@ protected function createDiscountOrderItemData(
 
         return $discountOrderItemData;
     }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Order\OrderData $orderData
+     * @param array $validProductIds
+     * @return \Shopsys\FrameworkBundle\Component\Money\Money
+     */
+    protected function calculateTotalApplicableProductsPrice(OrderData $orderData, array $validProductIds): Money
+    {
+        $totalPrice = Money::zero();
+
+        foreach ($orderData->items as $item) {
+            if (in_array($item->product->getId(), $validProductIds, true)) {
```

Comment:
in theory, the $items property may contain different item types than products (for example when someone changes the order of middlewares), and the product property may be null by definition. 

I suggest using 

```
foreach ($orderData->getItemsByType(OrderItemTypeEnum::TYPE_PRODUCT) as $item) {
    if (in_array($item->product?->getId(), $validProductIds, true)) {

Reply:
Understood and changed.